### ISSUE TITLE
Zedux v2 Roadmap

### ISSUE THREAD

**Issue by bowheart** [Omnistac/zedux#118]
Zedux v2 is on the horizon! Here's what we're planning on adding, deprecating, and removing plus some stuff we're still figuring out.

## Planned Deprecations

- `AtomInstance#getState` - replaced with `GraphNode#get` (which `AtomInstance` inherits).
- The `getInstance` atom getter - replaced with `getNode`.
- The `select` atom getter - replaced with `get`.
- `useAtomSelector` - replaced with `useAtomValue`.
- `injectAtomSelector` - replaced with `injectAtomValue` or the `get` atom getter.
- `AtomInstance#setStateDeep`. This has problems with `undefined` in TS (see #95). The `Store` class will retain its `setStateDeep` method but changed to make undefined values a no op. Both `AtomInstance` and `SignalInstance` will instead have `mutate` methods that finally introduce immer-style proxies natively to Zedux and fix the `undefined` problem naturally since they won't deal with any `RecursivePartial` types.
- `injectStore` - replaced with `injectSignal` (this will almost definitely not be part of v2 initial release, but a minor version after).
- The `Store` class and all helpers associated with it - `createStore`, `getMetaData`, `removeAllMeta`, `removeMeta`, all store-specific `zeduxTypes` (see #115), `actionFactory`, `createReducer`, and all type helpers for stores, actions, reducers, and subscribers.
- All `Atom*Type` type utils - replaced with `*Of` (e.g. `AtomStateType<myAtom>` -> `StateOf<myAtom>`).
- The `action` property on evaluation reasons.

## Planned Removals

- The `Selectors` class (see #114).
- The `Graph` class (see #114).
- The `EvaluationStack` class (see #114).
- The `AtomInstanceBase` class (see #114).
- The `SelectorCache` class (see #114) - replaced with `SelectorInstance`.
- `AtomInstance#addDependent` and `Selectors#addDependent` are replaced with `GraphNode#on` (see #114).
- The `IdGenerator` class - replaced with methods on `Ecosystem`.
- The `ZeduxPlugin` class and its actions - plugins are being completely reworked (see below).
- `internalStore` and `getEcosystem` - Zedux will no longer store ecosystems in its module-level state. This keeps it pure(r) and prevents GC surprises - if an ecosystem with no callbacks registered with the JS runtime goes completely out of scope, it and all its atoms should be able to be cleaned up.
- The `sourceType` property on evaluation reasons.
- `Ecosystem#registerPlugin` - replaced with `Ecosystem#plugin`.
- `injectInvalidate`. This is already deprecated, replaced by `injectSelf().invalidate`. V2 removes it.
- `manualHydration` atom config option - Zedux will instead auto-detect any usages of the `hydrate` option in `injectSignal`.
- The `AtomSelectorOrConfig` type - replaced with `SelectorTemplate`.
- The `AtomTemplateBase` class - renamed to `NodeTemplate`. `._createInstance` renamed to `.new` and `.getInstanceId` renamed to `.hash`.
- The `flags` atom config option - renamed to `tags`. This will affect the options passed to `Ecosystem#dehydrate` (and now `Ecosystem#findAll`) - `excludeFlags` renamed to `excludeTags` and `includeFlags` renamed to `includeTags`.
- The `store` property returned from `getInternals` - no direct replacement. Ecosystems will no longer be stored internally in Zedux module-level state. The `@@global` ecosystem will be returned from `createEcosystem({ id: '@@global' })`
- `getEcosystem` - removed with no replacement. For multi-window setups that leverage `getInternals`/`setInternals`, share your ecosystem reference across windows yourself rather than using `getEcosystem('my-id')` after calling `setInternals` in the child window. This should be about the same amount of code with less magic.
- `wipe` (the top-level exported function) - removed with no replacement. There's no internal store to clear.

## Planned Features

- The new Signal primitive (see #115)
- New plugin system - `ecosystem.on('change' | 'cycle' | 'edge' | 'error' | 'reset' | 'runEnd' | 'runStart', cb)`, `createEcosystem({ plugins })`, and `ecosystem.plugin(...plugins)` where `plugins` is an array of: `ecosystem => /* call ecosystem.on() */ MaybeCleanup`
- `api.set` to set all properties of an AtomAPI at once.
- Remove the string key requirement of `atom`. Keys will be auto-generated if not provided (but it's strongly recommended to pass one or use a build tool plugin to generate them for you).
- Events (see #115). In addition to the signal built-in events, atoms will have `.on('invalidate', cb)`. If we remove lifecycles from signals, atoms will also have `.on('cycle', cb)`.
- `optimize` option for `Ecosystem#batch` that will disable evaluation reason tracking for all updates resulting from any calls in the `.batch` callback.
- Automatic dependency tracking for `injectEffect`, `injectPromise`, `injectCallback`, and `injectMemo` - omit the deps array and Zedux will make the callback a reactive context similar to signals libs.
- Improved parameters to `Ecosystem#findAll`. Besides those already in #114, a `.findAll(callback)` overload for full control. That overload will apply to `Ecosystem#dehydrate` as well.

## Stuff We're Still Figuring Out

- Will phase 2 of the graph perf improvements be part of Zedux v2 or pushed back to v3?
- Is the lifecycle `status` property of graph nodes used enough to merit it getting a `status` getter on top of the obfuscated single-letter `l`ifecycleStatus property added by #114? We've never used it TMK for any purpose, but it's featured prominently in the docs. Maybe it just shouldn't be?
- Should we cut a `v2` of the docs and leave v1 alone? Or should we update the existing doc pages.
- Can the new signal primitive be made so it can use the [TC39 proposal](https://github.com/tc39/proposal-signals) internally when (if) it's finalized?
- The new type (introduced in #114) for arguments passed to `ecosystem.dehydrate` and `ecosystem.findAll` doesn't have a way to specify that only a certain type of node should be returned - atoms or selectors (or signals soon) or a custom node type. What would this argument look like?
- Instead of the new `ecosystem.live` atom getters, should we make the ecosystem's own atom getters reactive by default? We would then add `getOnce` and `getNodeOnce` methods which would have the current behavior of `get` and `getNode` respectively. This would get rid of the concept of atom getters and make the ecosystem all you need to know. It would also be able to have parity with signals - `signal.get` and `signal.getOnce`. It would also replace `injectAtomGetters` with `injectEcosystem` for better parity with Zedux's hooks.
- The timeline! When will Zedux v2 be available?

**Comment by bowheart** [comment]
The first release candidate has been published! https://github.com/Omnistac/zedux/releases/tag/v2.0.0-rc.0

Zedux v2 is now considered usable and stable, at least as far as all public, non-obfuscated APIs go. There may still be some "technically breaking" changes for internal APIs, but nothing normal users should need to know about.

There may be more features and bug fixes before v2.0.0 lands. That is the purpose of release candidates after all. And, of course, there's lots to document.

The [v2 migration guide](https://zedux.dev/docs/migrations/v2) can be considered a quick start for Zedux v2 for now.

TL;DR: If you don't mind the currently-sparse documentation, feel free to use v2!

```sh
pnpm i @zedux/react@next
```

**Comment by samuelhnrq** [comment]
Hey! Whats the status on this? v2 RC-10 seems to be stable for several months now, any updates on the timeline?

**Comment by bowheart** [comment]
Hey @samuelhnrq thanks for asking. The v2 RCs have indeed been stable for several months. We've been using it in prod since rc.9 I believe. The only thing preventing the full release is docs which I've been getting to here and there since April. Most of the API docs are done and PR'd.

I'm currently working on a big side project (with Zedux and TanStack Start btw) that's taking lots of my time, but I'm hoping to get back to the docs in maybe two weeks. If I can knock everything out, early November sounds about right for the official release. I don't feel I can commit to a specific date right now, but it might be even sooner if I free up 50 hours or so soon.

AI tooling really needs the new docs, but if you're not too dependent on that, I do recommend using v2 over v1 already. It's a big upgrade that is already very battle tested and the TypeScript, code comments, and PR'd docs should be enough to get by.
